File: runme/mcgen/local/datan/datsrc/regpol.c
#include <math.h>
#include <limits.h>
#include <stdlib.h>
#include "datsrc.h"

/* Common Block Declarations */

/*
struct {
    double g[1000];
} dasv04_;

#define dasv04_1 dasv04_
*/

void LIBFUNC regpol_(double const P_T *t, double const P_T *y,
             double const P_T *deltay, integer const P_T *n,
             integer const P_T *nr, double P_T *x, double P_T *b,
             double P_T *a, double P_T *chi2)
{
    /* System generated locals */
    integer b_dim1, b_offset, a_dim1, a_offset, i__1, i__2, i__3, i__4;
    double d__1;

    /* Local variables */
    integer i, j, k;
    double beta, tbar, s, alpha, gamma1, gamma2, sg;
    double *g;
    unsigned long amem;

    if ((amem = sizeof(double) * *n) > MAXALLOC) {
        MEMERROR("regpol_ (too large)");
        goto f_0;
    }
    g = (double *)MEMALLOC(amem);
    if (!g) {
        MEMERROR("regpol_");
        goto f_0;
    }

/* compute weights G and weighted mean TBAR */
    /* Parameter adjustments */
    --deltay;
    --y;
    --t;
    --chi2;
    a_dim1 = *n;
    a_offset = a_dim1 + 1;
    a -= a_offset;
    b_dim1 = *nr;
    b_offset = b_dim1 + 1;
    b -= b_offset;
    --x;

    /* Function Body */
    sg = 0.;
    tbar = 0.;
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
/* Computing 2nd power */
        d__1 = deltay[i];
        g[i - 1] = 1. / (d__1 * d__1);
        sg += g[i - 1];
        tbar += g[i - 1] * t[i];
    }
    tbar /= sg;
/* compute B and A for NR=1 */
    b[b_dim1 + 1] = 1. / sqrt(sg);
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        a[i + a_dim1] = b[b_dim1 + 1];
    }
/* compute B and A for NR=2 */
    if (*nr >= 2) {
        s = 0.;
        i__1 = *n;
        for (i = 1; i <= i__1; ++i) {
/* Computing 2nd power */
            d__1 = t[i] - tbar;
            s += g[i - 1] * (d__1 * d__1);
        }
        b[(b_dim1 << 1) + 2] = 1. / sqrt(s);
        b[b_dim1 + 2] = -b[(b_dim1 << 1) + 2] * tbar;
        i__1 = *n;
        for (i = 1; i <= i__1; ++i) {
            a[i + (a_dim1 << 1)] = b[b_dim1 + 2] + b[(b_dim1 << 1) + 2] * t[i];
        }
    }
/* compute B and A for NR greater than 2 */
    if (*nr > 2) {
        i__1 = *nr;
        for (j = 3; j <= i__1; ++j) {
            alpha = 0.;
            beta = 0.;
            gamma2 = 0.;
            i__2 = *n;
            for (i = 1; i <= i__2; ++i) {
/* Computing 2nd power */
                d__1 = a[i + (j - 1) * a_dim1];
                alpha += g[i - 1] * t[i] * (d__1 * d__1);
                beta += g[i - 1] * t[i] * a[i + (j - 1) * a_dim1] *
                        a[i + (j - 2) * a_dim1];
            }
            i__2 = *n;
            for (i = 1; i <= i__2; ++i) {
/* Computing 2nd power */
                d__1 = (t[i] - alpha) * a[i + (j - 1) * a_dim1] -
                       beta * a[i + (j - 2) * a_dim1];
                gamma2 += g[i - 1] * (d__1 * d__1);
            }
            gamma1 = 1. / sqrt(gamma2);
            b[j + b_dim1] = gamma1 * (-alpha * b[j - 1 + b_dim1] -
                            beta * b[j - 2 + b_dim1]);
            if (j >= 4) {
                i__2 = j - 2;
                for (k = 2; k <= i__2; ++k) {
                    b[j + k * b_dim1] = gamma1 * (b[j - 1 + (k - 1) * b_dim1]
                                        - alpha * b[j - 1 + k * b_dim1] -
                                        beta * b[j - 2 + k * b_dim1]);
                }
            }
            b[j + (j - 1) * b_dim1] = gamma1 * (b[j - 1 + (j - 2) * b_dim1] -
                    alpha * b[j - 1 + (j - 1) * b_dim1]);
            b[j + j * b_dim1] = gamma1 * b[j - 1 + (j - 1) * b_dim1];
            i__2 = *n;
            for (i = 1; i <= i__2; ++i) {
                a[i + j * a_dim1] = b[j + b_dim1];
                i__3 = j;
                for (k = 2; k <= i__3; ++k) {
                    i__4 = k - 1;
                    a[i + j * a_dim1] += b[j + k * b_dim1] *
                                         pow(t[i], (double)i__4);
                }
            }
        }
    }
/* compute X and CHI2 */
    i__1 = *nr;
    for (j = 1; j <= i__1; ++j) {
        x[j] = 0.;
        chi2[j] = 0.;
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
            x[j] += g[i - 1] * a[i + j * a_dim1] * y[i];
        }
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
            s = 0.;
            i__3 = j;
            for (k = 1; k <= i__3; ++k) {
                s += a[i + k * a_dim1] * x[k];
            }
/* Computing 2nd power */
            d__1 = y[i] - s;
            chi2[j] += g[i - 1] * (d__1 * d__1);
        }
    }
    MEMFREE((void *)g);
f_0:
    return;
} /* regpol_ */

File: runme/mcgen/local/datan/datsrc/regpol.f
      SUBROUTINE REGPOL(T,Y,DELTAY,N,NR,X,B,A,CHI2)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION T(N),Y(N),DELTAY(N),X(NR),CHI2(NR)
      DIMENSION B(NR,NR),A(N,NR)
      PARAMETER(MAXN=1000)
      DIMENSION G(MAXN)
      COMMON /DASV04/ G
      PARAMETER (ZERO=0.D0,ONE=1.D0)
C compute weights G and weighted mean TBAR
      SG=ZERO
      TBAR=ZERO
      DO 10 I=1,N
        G(I)=ONE/DELTAY(I)**2
        SG=SG+G(I)
        TBAR=TBAR+G(I)*T(I)
   10 CONTINUE
      TBAR=TBAR/SG
C compute B and A for NR=1
      B(1,1)=ONE/SQRT(SG)
      DO 20 I=1,N
        A(I,1)=B(1,1)
   20 CONTINUE
C compute B and A for NR=2
      IF(NR.GE.2) THEN
        S=ZERO
        DO 30 I=1,N
          S=S+G(I)*(T(I)-TBAR)**2
   30   CONTINUE
        B(2,2)=ONE/SQRT(S)
        B(2,1)=-B(2,2)*TBAR
        DO 40 I=1,N
          A(I,2)=B(2,1)+B(2,2)*T(I)
   40   CONTINUE
      END IF
C compute B and A for NR greater than 2
      IF(NR.GT.2) THEN
        DO 100 J=3,NR
          ALPHA=ZERO
          BETA=ZERO
          GAMMA2=ZERO
          DO 50 I=1,N
            ALPHA=ALPHA+G(I)*T(I)*A(I,J-1)**2
            BETA=BETA+G(I)*T(I)*A(I,J-1)*A(I,J-2)
   50     CONTINUE
          DO 60 I=1,N
            GAMMA2=GAMMA2+G(I)*((T(I)-ALPHA)*A(I,J-1)-
     +             BETA*A(I,J-2))**2
   60     CONTINUE
          GAMMA1=ONE/SQRT(GAMMA2)
          B(J,1)=GAMMA1*(-ALPHA*B(J-1,1)-BETA*B(J-2,1))
          IF(J.GE.4) THEN
            DO 70 K=2,J-2
              B(J,K)=GAMMA1*(B(J-1,K-1)-ALPHA*B(J-1,K)-
     +               BETA*B(J-2,K))
   70       CONTINUE
          END IF
          B(J,J-1)=GAMMA1*(B(J-1,J-2)-ALPHA*B(J-1,J-1))
          B(J,J)=GAMMA1*B(J-1,J-1)
          DO 90 I=1,N
            A(I,J)=B(J,1)
            DO 80 K=2,J
              A(I,J)=A(I,J)+B(J,K)*T(I)**(K-1)
   80       CONTINUE
   90     CONTINUE
  100   CONTINUE
      END IF
C compute X and CHI2
      DO 140 J=1,NR
        X(J)=ZERO
        CHI2(J)=ZERO
        DO 110 I=1,N
          X(J)=X(J)+G(I)*A(I,J)*Y(I)
  110   CONTINUE
        DO 130 I=1,N
          S=ZERO
          DO 120 K=1,J
            S=S+A(I,K)*X(K)
  120     CONTINUE
          CHI2(J)=CHI2(J)+G(I)*(Y(I)-S)**2
  130   CONTINUE
  140 CONTINUE
      END
